Pipeline benchmarking documentation update from UX feedback (#48)
intel-retail/automated-self-checkout: docs_src/pipelinebenchmarking.md

# Pipeline Benchmarking

> Pipeline benchmarking are done through collections of scripts to obtain the metrics of pipeline performance like video processing in frame-per-second (FPS),

Reviewer: ```suggestion
Pipeline benchmarking is done through a collections of scripts to obtain the pipeline performance metrics like video processing in frame-per-second (FPS),
```

> Pipeline benchmarking are done through collections of scripts to obtain the metrics of pipeline performance like video processing in frame-per-second (FPS),

Reviewer 2: Slight wording suggestion

Pipeline benchmarking is done through a collection of scripts to...

how much memory is used, how much power is consumed, ... and so on.

## Prerequisites: 

> Pipeline setup needs to be done first, pipeline setup documentation be found [HERE](./pipelinesetup.md).

Reviewer 3: ```suggestion
Pipeline setup needs to be completed first, pipeline setup documentation can be found [HERE](./pipelinesetup.md).
```

## Step 1: Build Benchmark Docker Images

> Benchmark scripts are containerized inside docker, depending on platforms/hardware you have, refer to the following table to choose one to build:

Reviewer: ```suggestion
Benchmark scripts are containerized inside Docker, depending on platforms/hardware you have, refer to the following table to choose one to build:
```

    Build command may take a while to run depending on your internet connection and machine specifications.

## Step 2: Run Benchmark
The `benchmark.sh` shell script is located under `benchmark_scripts` directory under the base directory.  Before executing this script,
change the current directory to directory `benchmark_scripts`.

This script will start benchmarking a specific number of pipelines or can start stream density benchmarking based on the desired FPS to reach.  

> Before one can run pipeline benchmarking for a given case, one has to determine several inputs listed as follows:

Reviewer: ```suggestion
Before running pipeline benchmarking for a given use case, determine the appropriate inputs, if any, from the list below:
```

### Determine the input source type

> The benchmark script can take one of different types from video input source as follows:

Reviewer: ```suggestion
The benchmark script can take one of these video input sources as described below:
```

### Real Time Streaming Protocol (RTSP)

    --inputsrc rtsp://127.0.0.1:8554/camera_0

> - **__NOTE:__** using RTSP source with the benchmark.sh will automatically run the camera simulator. The camera simulator will start an RTSP stream for each video file found in the `sample-media` folder.

Reviewer: is this the Note format you decided to use? just want to make sure everything is consistent.

Author: that's a good question: i think there is not bold type on the other  pages, i can remove the bold type to make it consistent with other pages

Author: oh and also use !!!

### USB Camera

>     --inputsrc /dev/video<N>, where N is 0 or an integer number

Reviewer: what does N represent - is it the number of cameras or the camera device ID?
```suggestion
    --inputsrc /dev/video<N>, where N is 0 or an integer
```

### RealSense Camera

    --inputsrc <RealSense camera serial number>

#### Obtaining RealSense camera serial number

[Follow this link to get serial number](./camera_serial_number.md)

### Video File

    --inputsrc file:my_video_file.mp4

- **__NOTE:__** video files must be in `sample-media` folder to be accessible from the Docker container. You can provide your own video files or download a sample video file using the script [download_sample_videos.sh](https://github.com/intel-retail/automated-self-checkout/blob/main/benchmark-scripts/download_sample_videos.sh).

---
### Determine the platform

#### Intel® Core

> - `--platform core.x` should be replaced with targeted GPUs such as core (for all GPUs), core.0, core.1, etc if GPUs are available

Reviewer 3: ```suggestion
- `--platform core.x` if GPUs are available, then replace this parameter with targeted GPUs such as core (for all GPUs), core.0, core.1, etc 
```

    Build command may take a while to run depending on your internet connection and machine specifications.

## Step 2: Run Benchmark
The `benchmark.sh` shell script is located under `benchmark_scripts` directory under the base directory.  Before executing this script,
change the current directory to directory `benchmark_scripts`.

This script will start benchmarking a specific number of pipelines or can start stream density benchmarking based on the desired FPS to reach.  
Before running pipeline benchmarking for a given use case, determine the appropriate inputs, from the list below:

### Determine the input source type

The benchmark script can take one of these video input sources as described below:

### Real Time Streaming Protocol (RTSP)

    --inputsrc rtsp://127.0.0.1:8554/camera_0

!!! note
    Using RTSP source with the benchmark.sh will automatically run the camera simulator. The camera simulator will start an RTSP stream for each video file found in the `sample-media` folder.

### USB Camera

    --inputsrc /dev/video<N>, where N is 0 or an integer

### RealSense Camera

    --inputsrc <RealSense camera serial number>

#### Obtaining RealSense camera serial number

> [Follow this link to get serial number](./camera_serial_number.md)

Reviewer 4: I think this would read better as "Follow this link to find how to get the serial number"

---

### Specified number of pipelines

One of main purposes to run benchmarking with specified number of pipelines is to discover the performance and system requirements for a given use case.

> An example to run benchmarking pipelines with specified number of pipelines:

Reviewer: ```suggestion
**Example:** run benchmarking pipelines with specified number of pipelines:
```

```bash
sudo ./benchmark.sh --pipelines <number of pipelines> --logdir <output dir>/data --init_duration 30 --duration 120 --platform <core|xeon|dgpu.x> --inputsrc <ex:4k rtsp stream with 10 objects>
```

where some of configurable input parameters are:

> --logdir is to configure the benchmarking output directory

Reviewer: ```suggestion
--logdir configures the benchmarking output directory
```

> --duration is to configure how long the benchmarking will run in number of seconds

Reviewer: ```suggestion
--duration configures how long the benchmarking will run in seconds
```

> --init_duration is to configure how long the initial time to wait before the benchmarking to take metrics or data collection in number of seconds

Reviewer: ```suggestion
--init_duration configures how long to wait for system initialization before the benchmarking metrics or data collection begins
```

Reviewer: don't have to use this wording exactly, but just cleaning it up a bit to make it more concise.

> and this pipeline benchmarking can be repetitively run for several different input sources and/or number of pipelines before consolidate

Reviewer: Is this supposed to be part of the `--init_duration description`? Where does this `and` begin?

```suggestion
and this pipeline benchmarking can be repetitively run for several different input sources and/or number of pipelines before consolidating
```

Reviewer 3: consolidating or consolidation ?

> and this pipeline benchmarking can be repetitively run for several different input sources and/or number of pipelines before consolidate

Reviewer 2: Slight wording suggestion

Multiple pipeline benchmarking runs with different configurations can be completed before the consolidation script is ran.

all pipeline output results.

To get consolidated pipeline results, run the following `make` command:
```bash
make consolidate ROOT_DIRECTORY=<output dir>
```
and this will give all the performance metrics among different workload cases given the same root directory specified by `ROOT_DIRECTORY` as shown above.

> One of consolidate example outputs is shown below:

Reviewer 3: ```suggestion
One of the consolidation example output is shown below:
```

11,GPU_0 GPU Utilization %,17.282
```

---
### Stream density

> Another pipeline benchmarking can do is to discover the maximum number of workloads/streams that can be ran in parallel for a given target FPS.  This can be useful to determine the hardware requirements in order to achieve the desired performance for input sources.

Reviewer 2: Another thing pipeline benchmarking...

```bash
sudo ./benchmark.sh --stream_density <target FPS> --logdir <output dir>/data --init_duration 30 --duration 120 --platform dgpu --inputsrc <ex:4k rtsp stream with 10 objects> --ocr 5 GPU
```

---
## Appendix: Benchmark Helper Scripts

- **camera-simulator.sh**

Starts the camera simulator. To use, place the script in a folder named camera-simulator. At the same directory level as the camera-simulator folder, create a folder called sample-media. The camera-simulator.sh script will start a simulator for each .mp4 video that it finds in the sample-media folder and will enumerate them as camera_0, camera_1 etc.  Be sure the path to camera-simulator.sh script is correct in the camera-simulator.sh script.  

- **stop_server.sh**

> Stops the docker images closing the pipelines  

Author: fixed
